ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -578,21 +596,29 @@ KEXT_STATIC int HandleVnodeOperation(
         {
             if (FileFlagsBitIsSet(currentVnodeFileFlags, FileFlags_IsEmpty))
             {
-                bool isWriteOperation = ActionBitIsSet(
-                    action,
-                    KAUTH_VNODE_WRITE_ATTRIBUTES |
-                    KAUTH_VNODE_WRITE_EXTATTRIBUTES |
-                    KAUTH_VNODE_WRITE_DATA |
-                    KAUTH_VNODE_APPEND_DATA);
+                // Prevent access to empty files in offline roots, except always allow the user to delete files.
+                bool shouldBlockIfOffline =
+                    (isRename && s_osSupportsRenameDetection) ||
+                    ActionBitIsSet(
+                        action,
+                        // Writes would get overwritten by subsequent hydration
+                        KAUTH_VNODE_WRITE_ATTRIBUTES |
+                        KAUTH_VNODE_WRITE_EXTATTRIBUTES |
+                        KAUTH_VNODE_WRITE_DATA |
+                        KAUTH_VNODE_APPEND_DATA |
+                        // Reads would yield bad (null) data
+                        KAUTH_VNODE_READ_DATA |
```

Comment:
This change *I think* will have us call Hydrate on for some new attributes.  Does that sound correct?

Reply:
That is certainly not the intention. The call to `ProviderMessaging_TrySendRequestAndWaitForResponse` below hasn't changed, the intended change here is to block reads from empty files when the provider is offline - of course precisely because it's not possible to hydrate in this case.